### PR TITLE
 added XdrivePoseEstimator to fuse measurements from the drive wheels, the odometry wheels, and the imu

### PR DESCRIPTION
**Pull request by benfeinberg**

### CODE REVIEW

File: 2021_CommandBased/src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -4,12 +4,18 @@
 
 package frc.robot;
 
+import edu.wpi.first.wpilibj.SerialPort;
```
Comment (paulbartell) [inline 2021_CommandBased/src/main/java/frc/robot/Constants.java:7]: The IMU is connected via I2C, not serial.

File: 2021_CommandBased/src/main/java/frc/robot/subsystems/Drive_s.java [hunk 2]
```diff
@@ -38,6 +41,10 @@
 
     //create odometry
     XdriveOdometry odometry;
+
+    private final XdrivePoseEstimator poseEstimator;
+
+    private AHRS imu;
```
Comment (paulbartell) [inline 2021_CommandBased/src/main/java/frc/robot/subsystems/Drive_s.java:47]: I would prefer if shared resources like motor controllers and IMUs were in the RobotContainer class.

File: 2021_CommandBased/src/main/java/frc/robot/util/KalmanFilterLatencyCompensator.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
+// Copyright (c) FIRST and other WPILib contributors.
```
Comment (paulbartell) [inline 2021_CommandBased/src/main/java/frc/robot/util/KalmanFilterLatencyCompensator.java:1]: How was this modified from the default version?

File: 2021_CommandBased/src/main/java/frc/robot/util/XdrivePoseEstimator.java [hunk 4]
```diff
@@ -0,0 +1,270 @@
+// Copyright (c) FIRST and other WPILib contributors.
```
Comment (paulbartell) [inline 2021_CommandBased/src/main/java/frc/robot/util/XdrivePoseEstimator.java:1]: Can you summarize the difference between this and the default PoseEstimators?

File: 2021_CommandBased/src/main/java/frc/robot/util/KalmanFilterLatencyCompensator.java (hunk 3 above)
Comment (benfeinberg) [reply 2021_CommandBased/src/main/java/frc/robot/util/KalmanFilterLatencyCompensator.java:1]: the constructor was made public

File: 2021_CommandBased/src/main/java/frc/robot/subsystems/Drive_s.java (hunk 2 above)
Comment (benfeinberg) [reply 2021_CommandBased/src/main/java/frc/robot/subsystems/Drive_s.java:47]: with command based robots, the convention is to put hardware in subsystems and subsystems in RobotContainer

File: 2021_CommandBased/src/main/java/frc/robot/util/XdrivePoseEstimator.java (hunk 4 above)
Comment (benfeinberg) [reply 2021_CommandBased/src/main/java/frc/robot/util/XdrivePoseEstimator.java:1]: the default PoseEstimators are all drivetrain specific, so this one works with the other xdrive classes

File: 2021_CommandBased/src/main/java/frc/robot/util/KalmanFilterLatencyCompensator.java (hunk 3 above)
Comment (paulbartell) [reply 2021_CommandBased/src/main/java/frc/robot/util/KalmanFilterLatencyCompensator.java:1]: Perhaps extending the included version of KalmanFilterLatencyCompensator and giving it a new name would be a better way to achieve this?

File: 2021_CommandBased/src/main/java/frc/robot/subsystems/Drive_s.java (hunk 2 above)
Comment (paulbartell) [reply 2021_CommandBased/src/main/java/frc/robot/subsystems/Drive_s.java:47]: After looking at the drive subsystem a bit more, I don't think there would be a great need to unit test since the logic is mostly in other objects.

File: 2021_CommandBased/src/main/java/frc/robot/util/KalmanFilterLatencyCompensator.java (hunk 3 above)
Comment (benfeinberg) [reply 2021_CommandBased/src/main/java/frc/robot/util/KalmanFilterLatencyCompensator.java:1]: I can't extend the wpilib version because it doesn't have a public constructor

File: 2021_CommandBased/src/main/java/frc/robot/util/XdrivePoseEstimator.java (hunk 4 above)
Comment (paulbartell) [reply 2021_CommandBased/src/main/java/frc/robot/util/XdrivePoseEstimator.java:1]: Could the kinematics class inherit from an existing kinematics class? Maybe that would remove the need for extra copies.

Comment (benfeinberg) [reply 2021_CommandBased/src/main/java/frc/robot/util/XdrivePoseEstimator.java:1]: That would probably be the right solution, but I didn't feel like redoing all of the classes that would need this. I also expect this class to need further changes, and it's easier to understand the code when it's all in one place.

**Comment by paulbartell** [comment]
I'm not sure removing the latency compensation is such a good idea. IMUs tend to lag significantly compared to encoders and this will confuse the UKF without some kind of compensation. The UKF internal to imu and I2C bus overhead are the main contributors here.

For this afternoon, let's publish a pose to the dashboard for each set of measurements so we can validate each separately:
- Pose estimate from drive wheel encoders + drive kinematics
- Pose estimate from odometry wheels + odometry wheel kinematics
- Direction estimate from IMU.
- (already there) Pose estimate from UKF

Additionally, if we could graph the measured velocity of each encoder (preferably on a single graph with a legend), that would be quite helpful for tuning the wheel velocity pid loops (and determining the correct feedforward value). Extra points if you can display this in both m/s and TalonFX native units (encoder counts per 100ms).
